refresh.template.py

```diff
@@ -90,6 +97,57 @@ def _get_headers(compile_args: typing.List[str], source_path_for_sanity_check: t
     return headers
 
 
+def _get_headers_msvc(compile_args: typing.List[str]):
+    # We don't want to compile, but only pre-process and show all the included header files.
+    # We don't want to have the preprocessed results written to a file, so we get them printed
+    # to `stderr`.
+    header_cmd = list(compile_args) + ["/showIncludes", "/EP"]
+
+    # cl.exe expects `INCLUDE` (and normally `LIB` but that is irrelevant here) to be set so that it can find the
+    # system headers. We have no idea what the correct paths are. So in case these variables are not set, we need
+    #  to infer them somehow. These variables might not be set if the user has a custom `cc_toolchain()`
+    # configured with `bazel` and does not rely on having sourced vcvarsall.bat.
+    environ = dict(os.environ)
+    if "INCLUDE" not in environ:
+        # We assume the base MSVC install based based on the `cl.exe` path in our command.
+        base = pathlib.Path(header_cmd[0]).parent.parent.parent.parent
+        # And we find the newest Windows 10 kits version that exists.
+        windows_kits = sorted( pathlib.Path("C:\\Program Files (x86)\\Windows Kits\\10\\include").glob("*"))[-1]
+        environ["INCLUDE"] =  os.pathsep.join(
+            (
+                os.fspath(base / "ATLMFC/include"),
+                os.fspath(base / "include"),
+                os.fspath(windows_kits / "ucrt"),
+                os.fspath(windows_kits / "shared"),
+                os.fspath(windows_kits / "um"),
+                os.fspath(windows_kits / "winrt"),
+                os.fspath(windows_kits / "cppwinrt"),
+            )
+        )
+
+    header_search_process  = subprocess.run(
+        header_cmd,
+        cwd=os.environ["BUILD_WORKSPACE_DIRECTORY"],
+        stderr=subprocess.PIPE,
+        stdout=subprocess.DEVNULL,
+        env=environ,
+        encoding="utf-8",
+        check=False, # We explicitly ignore errors and carry on.
+    )
+
+    if not header_search_process.stderr:
+        return [] # Worst case, we couldn't get the headers,
+
+    headers = (
+        # TODO This is not enough. Based on the locale, `cl.exe` will emit different marker strings. This should be extendend to multiple languages.
```

Comment:
Ugh, I'm so sorry MSVC does this.

Brainstorming some crazy ideas:
- We could try to call Bazel's code? (Not sure why they're extracting headers, but I remember that code you linked.)
- Is there a way to set the compiler's locale?
- We could invoke the compiler twice, with and without /showIncludes, and scan the lines that differ for the last instance of ": "?

Reply:
> We could try to call Bazel's code? (Not sure why they're extracting headers, but I remember that code you linked.)

That is Java code, so I guess not.

> Is there a way to set the compiler's locale?

No idea. Might depend on installed language packs for MSVC.

> We could invoke the compiler twice, with and without /showIncludes, and scan the lines that differ for the last instance of ": "?

Would increase generation time for the `compile_commands.json` quite drastically.

Reply:
> I did do some more digging, and maybe setting VSLANG=1033 would do the trick, getting everything always printed in english? (I saw https://github.com/ninja-build/ninja/issues/613) If it works, that would be so much better.

I tried setting it to turkish just for the header finding `cl.exe` call using this trick and it did not work. A comment says "before calling vcvars32/64" and that is out of our control.

Reply:
> If that fails, at a minimum, could we match the language prefixes that Bazel does and you linked to? (Languages also listed, maybe more easily, in that ninja issue.

Done.